### PR TITLE
Pull Request for Issue1142: Add Internal Retry to AMControlMoveAction

### PR DESCRIPTION
**Pull request by davidChevrier**
Added and tested on VESPERS. Appears to work well. Default retries is set to 1.

### CODE REVIEW

File: source/actions3/actions/AMControlMoveAction3.cpp [hunk 1]
```diff
@@ -44,6 +47,32 @@ AMControlMoveAction3::AMControlMoveAction3(const AMControlMoveAction3 &other)
 		control_ = AMBeamline::bl()->exposedControlByInfo(*(info->controlInfo()));
 	else
 		control_ = 0;
+
+	retries_ = 1;
```
Comment (dretrex) [inline source/actions3/actions/AMControlMoveAction3.cpp]: Do we not want to set the number of retries to the ```other.retries()```?  It is the copy constructor.

File: source/actions3/actions/AMControlMoveAction3.cpp [hunk 2]
```diff
@@ -85,15 +114,21 @@ void AMControlMoveAction3::startImplementation()
...
 	startPosition_ = control_->toInfo();
 
-	// start the move:
-	int failureExplanation;
+	attempts_ = 1;
+
+	// determine the absolute destination
 	if(controlMoveInfo()->isRelativeMove() && controlMoveInfo()->isRelativeFromSetpoint())
-		failureExplanation = control_->moveRelative(setpoint.value(), AMControl::RelativeMoveFromSetpoint);
+		destination_ = control_->setpoint() + setpoint.value();
```
Comment (dretrex) [inline source/actions3/actions/AMControlMoveAction3.cpp]: You did this change after I left, right?  Does this still work properly if a control move action was inside of a loop doing a relative move 10 times?

File: source/actions3/actions/AMControlMoveAction3.cpp [hunk 3]
```diff
@@ -151,10 +188,32 @@ void AMControlMoveAction3::onMoveFailed(int reason)
...
+	QString fundamentalFailureMessage = QString("There was an error moving the control '%1' into position (Retry %2, attempt %3 of %4). Target: %5 %6. Actual position: %7 %8. Tolerance %9. Reason: %10.").arg(control_->name()).arg(retryAvailableString).arg(attempts_).arg(retries_+1).arg(control_->setpoint()).arg(control_->units()).arg(control_->value()).arg(control_->units()).arg(control_->tolerance()).arg(AMControl::failureExplanation(reason));
 	// error message with reason
 	AMErrorMon::alert(this, definedFailureReason, QString("%1. Please report this problem to the beamline staff.").arg(fundamentalFailureMessage));
-	setFailed(fundamentalFailureMessage);
+
+	if(retryAvailable){
+		attempts_++;
+
+		// retry move
+		qDebug() << "Retrying move to " << destination_;
```
Comment (dretrex) [inline source/actions3/actions/AMControlMoveAction3.cpp]: Do you still want this qDebug() there?  Do you want to have an error monitor instead or is the error mon stuff handled entirely by the above error mon?

File: source/actions3/actions/AMControlMoveAction3.cpp (hunk 1 above)
Comment (davidChevrier) [inline source/actions3/actions/AMControlMoveAction3.cpp]: Clever @dretrex

File: source/actions3/actions/AMControlMoveAction3.cpp (hunk 2 above)
Comment (davidChevrier) [inline source/actions3/actions/AMControlMoveAction3.cpp]: Nope, this is what we tested yesterday. Seemed to work correctly as far as we could tell.

File: source/actions3/actions/AMControlMoveAction3.cpp (hunk 3 above)
Comment (davidChevrier) [inline source/actions3/actions/AMControlMoveAction3.cpp]: Gone.

**Pull request merged**